[govee] New Govee LAN-API Binding (#15696)
openhab/openhab-addons: bundles/org.openhab.binding.goveelan/README.md

- H7041 LED Outdoor Bulb String Lights
- H7042 LED Outdoor Bulb String Lights
- H705A Permanent Outdoor Lights 30M
- H705B Permanent Outdoor Lights 15M
- H7050 Outdoor Ground Lights 11M
- H7051 Outdoor Ground Lights 15M
- H7055 Pathway Light
- H7060 LED Flood Lights (2-Pack)
- H7061 LED Flood Lights (4-Pack)
- H7062 LED Flood Lights (6-Pack)
- H7065 Outdoor Spot Lights
- H6051 Aura - Smart Table Lamp
- H6056 H6056 Flow Plus
- H6059 RGBWW Night Light for Kids
- H618F RGBIC LED Strip Lights
- H618E LED Strip Lights 22m
- H6168 TV LED Backlight

## Discovery

Discovery is done by scanning the devices in the Thing section. 

The devices _do not_ support the LAN API support out-of-the-box.
To be able to use the device with the LAN API, the following needs to be done (also see the "Preparations for LAN API Control" section in the [Goveee LAN API Manual](https://app-h5.govee.com/user-manual/wlan-guide)):

+ Start the Govee APP and add / discover the device (via Bluetooth) as described by the vendor manual
Go to the settings page of the device 
![govee device settings](doc/device-settings.png)
+ Note that it may take several(!) minutes until this setting comes up.

> + Switch on the LAN-Control setting.

Reviewer: ```suggestion
+ Switch on the LAN Control setting.
```

at least according to the screenshot. 🙂

Author: `I guess I didn't manage to convince you about the prefixes.`

You did. I changed it:

error.could-not-query-device = Could not control/query device at IP address {0}
error.offline.ip-address.missing = Handler for thing could not be added to list because ipaddress == null
error.offline.communication-error.empty-response = Offline due to receiving an empty response

`I'm sorry, I should have suggested COMMUNICATION_ERROR here.``

Like so?

<img width="816" alt="image" src="https://github.com/openhab/openhab-addons/assets/5937600/49d23641-285d-48fb-81a5-5ee8d1e8e282">

To be honest, regarding the Singleton-topic, I would leave it for the time being - I have worked several times on concurrency (even weeks on the nanoleaf binding to find out the root cause of a reoccurring issue). I definitely works and it took me hours to make it work safely. If I would go down a complete new path, this would sure put the whole binding on-hold for a number of weeks. Hence, I would ask you not to insist to change something here for the time being while I am happy to look into further after the release of the binding.

Reviewer: > To be honest, regarding the Singleton-topic, I would leave it for the time being - I have worked several times on concurrency (even weeks on the nanoleaf binding to find out the root cause of a reoccurring issue). I definitely works and it took me hours to make it work safely. If I would go down a complete new path, this would sure put the whole binding on-hold for a number of weeks. Hence, I would ask you not to insist to change something here for the time being while I am happy to look into further after the release of the binding.

I'm sorry, but I will have to insist on this, since this will not only affect the binding, but can make the openHAB installation as a whole unstable. This is not easy to test manually, so even if your tests are successful, this is something that will occur randomly depending on timing etc.

Usually new bindings can take months to get reviewed, so believe me, you already have a head start here. 😉

+ Now the device can be used with openHAB.
+ The easiest way is then to scan the devices via the SCAN button in the thing section of that binding

## Thing Configuration

Even though binding configuration is supported via a thing file it should be noted that the IP address is required and there is no easy way to find out the IP address of the device.

> One possibility is to look for the MAC address in the Govee app and then looking the IP address up via 

Reviewer: ```suggestion
One possibility is to look for the MAC address in the Govee app and then looking the IP address up via:
```

```
arp -a | grep "MAC_ADDRESS"
```

| Name            | Type    | Description                           | Default | Required | Advanced |
|-----------------|---------|---------------------------------------|---------|----------|----------|
| hostname        | text    | Hostname or IP address of the device  | N/A     | yes      | no       |
| macAddress      | text    | MAC address of the device             | N/A     | yes      | no       |
| deviceType      | text    | The product number of the device      | N/A     | yes      | no       |

> | refreshInterval | integer | Interval the device is polled in sec. | 3       | no       | yes      |

Reviewer: ```suggestion
| refreshInterval | integer | Interval the device is polled in sec. | 5       | no       | yes      |
```

## Channels

| Channel             | Type              | Read/Write | Description         |
|---------------------|-------------------|------------|---------------------|
| brightness          | Percentage        | RW         |                     |
|                     | Switch            | RW         | Power On / OFF      |
| color               | Color HSB Type    | RW         |                     |
| colorTemperatureAbs | Color Temperature | RW         | in 2000-9000 Kelvin |

> Note: you may have to add "%.0f K" as the state description when creating a colorTemperatureAbs item.

Reviewer: ```suggestion
Note: you may have to add "%.0f K" as the state description when creating a color-temperature-abs item.
```

- H6168 TV LED Backlight

## Discovery

Discovery is done by scanning the devices in the Thing section. 

The devices _do not_ support the LAN API support out-of-the-box.
To be able to use the device with the LAN API, the following needs to be done (also see the "Preparations for LAN API Control" section in the [Goveee LAN API Manual](https://app-h5.govee.com/user-manual/wlan-guide)):

+ Start the Govee APP and add / discover the device (via Bluetooth) as described by the vendor manual
Go to the settings page of the device 
![govee device settings](doc/device-settings.png)
+ Note that it may take several(!) minutes until this setting comes up
+ Switch on the LAN-Control setting
+ Now the device can be used with openHAB
+ The easiest way is then to scan the devices via the SCAN button in the thing section of that binding

## Thing Configuration

Even though binding configuration is supported via a thing file it should be noted that the IP address is required and there is no easy way to find out the IP address of the device.
One possibility is to look for the MAC address in the Govee app and then looking the IP address up via 

```
arp -a | grep "MAC_ADDRESS"
```

> ### Thing Configuration

Reviewer: This chapter has the same name as the parent chapter, probably not intended?

Author: No, it's obsolete

# Govee Lan-API Binding

> ![govee](doc/govee.png)

Author: I know ... I just wanted at least one nice pic in the Readme during the initial contribution ;-)

> ![govee](doc/govee.png)

Reviewer: The link is still there.

> ![govee](doc/govee.png)

Author: Are you refering to

> This binding integrates Light devices from [Govee](https://www.govee.com/).

if not, the picture is not meant to be the Govee logo but a nice header where people can see what the binding supports. Similarly like in https://www.openhab.org/addons/bindings/avmfritz/

> ![govee](doc/govee.png)

Reviewer: You are right, sorry.

- H610B Music Wall Lights
- H6172 Outdoor LED Strip 10m
- H61B2 RGBIC Neon TV Backlight
- H61E1 LED Strip Light M1
- H7012 Warm White Outdoor String Lights
- H7013 Warm White Outdoor String Lights
- H7021 RGBIC Warm White Smart Outdoor String
- H7028 Lynx Dream LED-Bulb String
- H7041 LED Outdoor Bulb String Lights
- H7042 LED Outdoor Bulb String Lights
- H705A Permanent Outdoor Lights 30M
- H705B Permanent Outdoor Lights 15M
- H7050 Outdoor Ground Lights 11M
- H7051 Outdoor Ground Lights 15M
- H7055 Pathway Light
- H7060 LED Flood Lights (2-Pack)
- H7061 LED Flood Lights (4-Pack)
- H7062 LED Flood Lights (6-Pack)
- H7065 Outdoor Spot Lights
- H6051 Aura - Smart Table Lamp
- H6056 H6056 Flow Plus
- H6059 RGBWW Night Light for Kids
- H618F RGBIC LED Strip Lights
- H618E LED Strip Lights 22m
- H6168 TV LED Backlight

## Discovery

Discovery is done by scanning the devices in the Thing section. 

The devices _do not_ support the LAN API support out-of-the-box.
To be able to use the device with the LAN API, the following needs to be done (also see the "Preparations for LAN API Control" section in the [Goveee LAN API Manual](https://app-h5.govee.com/user-manual/wlan-guide)):

> - Start the Govee APP and add / discover the device (via bluetooth) as described by the vendor manual

Reviewer: ```suggestion
- Start the Govee APP and add / discover the device (via Bluetooth) as described by the vendor manual
```

- Go to the settings page of the device 
![govee device settings](doc/device-settings.png)

> - Note that it may take several(!) minutes until this setting comes up
> - Switch on the LAN-Control setting
> - Now the device can be used with openHAB
> - The easiest way is then to scan the devices via the SCAN button in the thing section of that binding

Reviewer: ```suggestion
- Note that it may take several(!) minutes until this setting comes up.
- Switch on the LAN-Control setting.
- Now the device can be used with openHAB.
- The easiest way is then to scan the devices via the SCAN button in the thing section of that binding.
```

Author: And that's me being the maintainer of the docs - was probably a bit late when I wrote that 🤦🏻‍♂️

## Binding Configuration

There is no particular binding configuration needed.

## Thing Configuration

Even though binding configuration is supported via a thing file it should be noted that the IP address is required and there is no easy way to find out the IP address of the device.
One possibility is to look for the MAC address in the Govee app and then looking the IP address up via 

> `arp -a | grep "MAC_ADDRESS"`

Author: Changed it and removed the backticks from the line itself

> `arp -a | grep "MAC_ADDRESS"`

Reviewer: Oh, right, I missed that.

> `arp -a | grep "MAC_ADDRESS"`

Reviewer: Can you add "shell" as language as well?

> ### `sample` Thing Configuration

Reviewer: I see this is now a duplicate of a new comment I made regarding duplicated chapter names. I would still suggest the above to prepare for more thing types.
